asyncscope.md

```diff
@@ -279,201 +352,541 @@ listening activity is bound in the scope of the loop, the lifetime of handling r
+scope", which happens when the sender is destroyed, its _`operation-state`_ is destroyed, or its _`operation-state`_ is
+completed. Moving a _`nest-sender`_ transfers responsibility for decrementing the count from the old instance to the new
+one. Copying a _`nest-sender`_ is permitted if the sender it's wrapping is copyable, but the copy may "fail" since
+copying requires incrementing the scope's count, which is only allowed when the scope is open; if copying fails, the new
+sender is an unassociated sender that behaves as if it were the result of a failed call to `nest()`.
+
+While a scope is closed or joined, calls to `nest(snd, scope)` will always fail by discarding the given sender and
+returning an unassociated _`nest-sender`_. Failed calls to `nest()` do not change the scope's count. Unassociated
+_`nest-senders`_ do not have a reference to the scope they came from and always complete with `stopped` when connected
+and started. Copying or moving an unassociated sender produces another unassociated sender.
+
+The state transitions of a `counting_scope` mean that it can be used to protect asynchronous work from use-after-free
+errors. Given a resource, `res`, and a `counting_scope`, `scope`, obeying the following policy is enough to ensure that
+there are no attempts to use `res` after its lifetime ends:
+
+- all senders that refer to `res` are nested within `scope`; and
+- `scope` is destroyed (and therefore joined) before `res` is destroyed.
+
+Under the standard assumption that the arguments to `nest()` are and remain valid while evaluating `nest()`, it is
+always safe to invoke any supported operation on the returned _`nest-sender`_. Furthermore, if all senders returned from
+`nest()` are eventually started or discarded then the `join()` operation always eventually finishes because the number
+of outstanding senders nested within the corresponding scope is monotonically decreasing. Conversely, "leaking" a
+_`nest-sender`_ will cause the `join()` operation to deadlock.
+
+The risk of deadlock is explicitly preferred in this design over the risk of use-after-free errors because
+`counting_scope` is an async scope that is biased towards being used to progressively add structure to
+generally-unstructured code. We've found that the central problem in progressively structuring unstructured code is
+determining appropriate bounds for each asynchronous task when those bounds are not clear; it is easier to figure out
+where to synchronously `join()` a scope than it is to ensure that all `spawn()`ed work is properly scoped within any
+particular object's lifetime. So, although it is generally easier to diagnose use-after-free errors than it is to
+diagnose deadlocks, we've found that it's easier to *avoid* deadlocks with this design than it is to avoid
+use-after-free errors with other designs.
+
+A `counting_scope` is uncopyable and immovable so its copy and move operators are explicitly deleted.
```

Comment:
A `counting_scope` _could_ be movable, but I think that would require an allocation. Probably it would be best to justify this design decision.